netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,177 @@
+        }
+
+        Ok(())
+    }
+
+    $(#[$meta])*
+    pub async fn [<poll_ $state>](&mut self) -> RussulaResult<Poll<()>> {
+        // Poll each peer protocol instance.
+        //
+        // If the peer is already in the desired state then this should be a noop.
+        for peer in self.instance_list.iter_mut() {
+            if let Err(err) = peer.protocol.[<poll_ $state>](&mut peer.stream).await {
+                if err.is_fatal() {
+                    error!("{} {}", err, peer.addr);
+                    panic!("{} {}", err, peer.addr);
+                }
+            }
+        }
+
+        // Check that all instances are at the desired state.
+        let poll = if self.[<is_ $state _state>]() {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        };
+        Ok(poll)
+    }
+
+    /// Check if all instances are at the desired state
+    fn [< is_ $state _state>](&self) -> bool {
+        for peer in self.instance_list.iter() {
+            // All instance must be at the desired state
+            if !peer.protocol.[< is_ $state _state>]() {
+                return false;
+            }
+        }
+        true
+    }
+}};
+}
+
+impl<P: Protocol + Send> Russula<P> {
+    state_api!(
+        /// Successfully connected to peer instances and ready to make progress.
+        ready
+    );
+    state_api!(
+        /// The coordination completed and reached a terminal state.
+        done
+    );
+    state_api!(
+        /// Note: Should only be called by Coordinators
+        ///
+        /// A intermediate state to query if Worker peers are in some
+        /// desired state.
+        worker_running
+    );
+}
+
+type SockProtocol<P> = (SocketAddr, P);
+pub struct RussulaBuilder<P: Protocol> {
+    /// Address on which the Coordinator and Worker communicate on.
+    ///
+    /// The Coordinator gets a list of workers addrs to 'connect' to. This can
+    /// be of size >= 1. The Worker gets its own addr to 'listen' on and should
+    /// be size = 1.
+    // TODO Create different Russula struct for Coordinator/Workers to capture
+    // different usage patterns.
+    protocol_addr_pair_list: Vec<SockProtocol<P>>,
+    poll_delay: Duration,
+}
+
+impl<P: Protocol> RussulaBuilder<P> {
+    pub fn new(peer_addr: BTreeSet<SocketAddr>, protocol: P, poll_delay: Duration) -> Self {
+        // TODO if worker check that the list is len 1 and points to local addr on which to listen
+        let mut peer_list = Vec::new();
+        peer_addr.into_iter().for_each(|addr| {
+            peer_list.push((addr, protocol.clone()));
+        });
+        Self {
+            protocol_addr_pair_list: peer_list,
+            poll_delay,
+        }
+    }
+
+    pub async fn build(self) -> RussulaResult<Russula<P>> {
+        let mut stream_protocol_list = Vec::new();
+        for (addr, protocol) in self.protocol_addr_pair_list.into_iter() {
+            let mut retry_attempts = 10;
```

Comment:
Nit: magic number, could we use a constant instead? I think I saw a "3 for package loss" some other place too.